### PR TITLE
Implement DSIGN for Ed25519

### PR DESCRIPTION
**Pull request by nc6**

### CODE REVIEW

File: cardano-crypto-class/src/Cardano/Crypto/DSIGN/Ed25519.hs [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        let vk = toPublic sk
+            bs = toStrict $ serialize a
+        return $ SigEd25519DSIGN $ sign sk vk bs
+
+    verifyDSIGN () (VerKeyEd25519DSIGN vk) a (SigEd25519DSIGN sig) =
+        if verify vk (toStrict $ serialize a) sig
+          then Right ()
+          else Left "Verification failed"
+
+    -- | Ed21159 key size is 32 octets (per https://tools.ietf.org/html/rfc8032#section-5.1.6)
```
Comment (intricate) [inline cardano-crypto-class/src/Cardano/Crypto/DSIGN/Ed25519.hs:74]:
```suggestion
    -- | Ed25519 key size is 32 octets (per https://tools.ietf.org/html/rfc8032#section-5.1.6)
```

File: shell.nix [hunk 2]
```diff
@@ -36,7 +36,7 @@ let
 
     # Prevents cabal from choosing alternate plans, so that
     # *all* dependencies are provided by Nix.
-    exactDeps = true;
+    exactDeps = false;
```
Comment (intricate) [inline shell.nix:39]: Is this change supposed to be a part of this PR? I don't really know anything about nix, just making sure that this wasn't a mistake.

Comment (dcoutts) [inline shell.nix:39]: I'm not complaining, but this is unrelated. Preferably in a separate patch.

File: cardano-crypto-class/src/Cardano/Crypto/DSIGN/Ed25519.hs [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+import Data.ByteArray (ByteArrayAccess, convert)
+import Data.ByteString (ByteString)
+import Data.ByteString.Lazy (toStrict)
+import GHC.Generics (Generic)
+
+data Ed25519DSIGN
+
+instance DSIGNAlgorithm Ed25519DSIGN where
+
+    type Signable Ed25519DSIGN = ToCBOR
```
Comment (dcoutts) [inline cardano-crypto-class/src/Cardano/Crypto/DSIGN/Ed25519.hs:38]: This is not right for the real thing. We must never re-serialise for verification. Instead this should be the class to recover the annotated bytes that we already have.

File: cardano-crypto-class/src/Cardano/Crypto/DSIGN/Ed25519.hs [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+    deriveVerKeyDSIGN (SignKeyEd25519DSIGN sk) = VerKeyEd25519DSIGN $ toPublic sk
+
+    signDSIGN () a (SignKeyEd25519DSIGN sk) = do
+        let vk = toPublic sk
+            bs = toStrict $ serialize a
+        return $ SigEd25519DSIGN $ sign sk vk bs
+
+    verifyDSIGN () (VerKeyEd25519DSIGN vk) a (SigEd25519DSIGN sig) =
+        if verify vk (toStrict $ serialize a) sig
```
Comment (dcoutts) [inline cardano-crypto-class/src/Cardano/Crypto/DSIGN/Ed25519.hs:70]: And then instead of `serialise` here is where we should recover the annotated bytes for the serialised rep.

Similarly for `serialise` in `signDSIGN`.

File: cardano-crypto-class/src/Cardano/Crypto/DSIGN/Ed25519.hs (hunk 3 above)
Comment (nc6) [reply cardano-crypto-class/src/Cardano/Crypto/DSIGN/Ed25519.hs:38]: The new Shelley annotations work in this way. The `ToCBOR` instance does _not_ serialise but instead uses the annotation

File: cardano-crypto-class/src/Cardano/Crypto/DSIGN/Ed25519.hs (hunk 4 above)
Comment (nc6) [reply cardano-crypto-class/src/Cardano/Crypto/DSIGN/Ed25519.hs:70]: As per previous comment: the annotated instances in Shelley provide that `serialise` does _not_ result in re-serialisation.

**Comment by nc6** [comment]
@dcoutts See for example https://github.com/input-output-hk/cardano-ledger-specs/blob/master/shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Tx.hs#L180

**Comment by dcoutts** [comment]
#101 covers the suggestion to make it much harder to make a mistake with re-serialising. So this PR is fine, and it can be improved with the idea from #101.
